File: Algorithm-java/src/BOJ/Silver/Q2583.java
package BOJ.Silver;

import java.io.*;
import java.util.*;

public class Q2583 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] input = br.readLine().trim().split(" ");
        int m = Integer.parseInt(input[0]);
        int n = Integer.parseInt(input[1]);
        int k = Integer.parseInt(input[2]);

        int[][] squares = new int[k][4];

        for (int i = 0; i < k; i++) {
            input = br.readLine().trim().split(" ");
            int[] sqare = {Integer.parseInt(input[0]), Integer.parseInt(input[1]), Integer.parseInt(input[2]), Integer.parseInt(input[3])};
            squares[i] = sqare;
        }

        int[][] board = new int[m][n];
        List<Integer> rst = new ArrayList<>();

        for (int i = 0; i < k; i++) {
            int[] square = squares[i];
            for (int j = square[1]; j < square[3]; j++) {
                for (int l = square[0]; l < square[2]; l++) {
                    board[j][l] = 1;
                }
            }
        }

        int[][] chk = new int[m][n];
        int[] dx = {0, -1, 0, 1};
        int[] dy = {1, 0, -1, 0};

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 0 && chk[i][j] == 0) {
                    Deque queue = new ArrayDeque();
                    queue.offer(new int[]{i, j});
                    int cnt = 0;
                    while (!queue.isEmpty()) {
                        int[] pos = (int[]) queue.poll();
                        int y = pos[0];
                        int x = pos[1];
                        if (chk[y][x] == 0) {
                            chk[y][x] = 1;
                            cnt++;

                            for (int l = 0; l < 4; l++) {
                                int nx = x + dx[l];
                                int ny = y + dy[l];

                                if (nx >= 0 && nx < n && ny >= 0 && ny < m && board[ny][nx] == 0 && chk[ny][nx] == 0) {
                                    queue.offer(new int[]{ny, nx});
                                }
                            }
                        }
                    }
                    rst.add(cnt);
                }
            }
        }

        Collections.sort(rst);
        bw.write(String.valueOf(rst.size()) + "\n");
        for (int num: rst) {
            bw.write(String.valueOf(num) + " ");
        }
        bw.flush();
    }
}

File: BOJ/BFS/Q2583.py

from collections import deque

M, N, K = map(int, input().split())
squares = [list(map(int, input().split())) for _ in range(K)]

board = [[0 for _ in range(N)] for _ in range(M)]
chk = [[0 for _ in range(N)] for _ in range(M)]
rst = []

for x1, y1, x2, y2 in squares:
    for y in range(y1, y2):
        for x in range(x1, x2):
            board[y][x] = 1

dy = [0, 1, 0, -1]
dx = [1, 0, -1, 0]

for i in range(M):
    for j in range(N):
        if board[i][j] == 0 and chk[i][j] == 0:
            queue = deque([[i, j]])
            cnt = 0
            while queue:
                y, x = queue.popleft()

                if board[y][x] == 0 and chk[y][x] == 0:
                    cnt += 1
                    chk[y][x] = 1

                    for d in range(4):
                        ny = y + dy[d]
                        nx = x + dx[d]

                        if 0 <= ny < M and 0 <= nx < N and board[ny][nx] == 0 and chk[ny][nx] == 0:
                            queue.append([ny, nx])

            rst.append(cnt)

rst.sort()
print(len(rst))
print(*rst)





